server/src/main.rs

```diff
@@ -0,0 +1,201 @@
+    ) -> tower_lsp::jsonrpc::Result<Option<GotoDefinitionResponse>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+
+        if let Some(location) = get_definition_span(&state, uri, position) {
+            let start = Position {
+                line: (location.start.row - 1) as u32,
+                character: (location.start.col - 1) as u32,
+            };
+
+            let end = Position {
+                line: (location.end.row - 1) as u32,
+                character: (location.end.col - 1) as u32,
+            };
+
+            Ok(Some(GotoDefinitionResponse::Scalar(Location {
+                uri: Url::from_file_path(location.file).unwrap(),
+                range: Range::new(start, end),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn hover(&self, params: HoverParams) -> tower_lsp::jsonrpc::Result<Option<Hover>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+        if let Some(info) = get_hover_info(&state, uri, position) {
+            Ok(Some(Hover {
+                contents: HoverContents::Scalar(MarkedString::String(info)),
+                range: None,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file opened!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+
+    async fn did_save(&self, params: DidSaveTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+}
+
+impl Backend {
+    async fn handle_file_change(&self, uri: Url) {
+        let (updated_state, options) = self.compile_slice_files().await;
+        let mut compilation_state_guard = self.compilation_state.lock().await;
+        *compilation_state_guard = updated_state;
+
+        let mut compilation_options_guard = self.compilation_options.lock().await;
+        *compilation_options_guard = options;
+
+        let diagnostics = self
+            .get_diagnostics_for_uri(
+                &uri,
+                &mut compilation_state_guard,
+                &compilation_options_guard,
+            )
+            .await;
+
+        self.client
+            .publish_diagnostics(uri, diagnostics, None)
+            .await;
+    }
+
+    // This will consume all of the diagnostics in the compilation state and return them as LSP diagnostics.
+    // If you need the diagnostics again you will need to recompile.
+    async fn get_diagnostics_for_uri(
```

Comment:
@InsertCreativityHere Is there a way for us to not consume the diagnostics?

Reply:
Not with the current API. The Diagnostics can get pretty heavy, so we always re-use them, instead of making pro-active clones of them.

Additionally, the `CompilationState` holds a set of unfixed diagnostics.
As compilation proceeds, we push diagnostics into it, but we don't fix them until the very end.
(By fix, I mean update the diagnostic levels based on any attributes, or CLI options).

I think it's fine for this use case anyways. we'd probably want a fresh set of diagnostics to report to VSCode, instead of reporting the diagnostics that we've already reported to it.